tests/unit/test_intervals.py

```diff
@@ -0,0 +1,251 @@
+"""
+Random data based tests for the interval iterator code.
+"""
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import unittest
+import random
+
+import ga4gh.backend as backend
+
+
+def setUp():
+    # set the random seed to this is reproducable.
+    random.seed(1)
+
+
+def intervalsIntersect(a1, b1, a2, b2):
+    """
+    Returns True if the specified half-closed intervals [a1, b1)
+    and [a2, b2) intersect.
+    """
+    # Intervals do not interset if [a1, b1) is wholly to the left
+    # or right of [a2, b2).
+    return not (b1 <= a2 or a1 >= b2)
+
+
+class TestIntervalsIntersect(unittest.TestCase):
+    """
+    Tests to ensure our interval intersection code works.
+    """
+    def testEmptyIntervals(self):
+        self.assertFalse(intervalsIntersect(0, 0, 0, 0))
+        self.assertFalse(intervalsIntersect(1, 1, 1, 1))
+        self.assertFalse(intervalsIntersect(1, 1, 1, 2))
+        self.assertFalse(intervalsIntersect(1, 10, 1, 1))
+
+    def testNonIntersecting(self):
+        tests = [
+            [0, 1, 1, 2],
+            [0, 1, 2, 3],
+            [2, 1, 0, 1],
+            [2, 3, 0, 1]]
+        for test in tests:
+            self.assertFalse(intervalsIntersect(*test))
+
+    def testIntersecting(self):
+        tests = [
+            [0, 1, 0, 1],
+            [0, 1, 0, 3],
+            [0, 3, 0, 1],
+            [1, 10, 9, 10],
+            [1, 10, 9, 12],
+            [1, 10, 5, 6],
+            [1, 10, 1, 2]]
+        for test in tests:
+            self.assertTrue(intervalsIntersect(*test))
+
+
+def randomIntervals(start, end, numIntervals):
+    """
+    Returns a list of numIntervals half-closed intervals. Each element
+    of the returned array is a tuple a, b such that a <= start < b <= end.
+    """
+    intervals = []
+    for _ in range(numIntervals):
+        a = random.randrange(start, end - 1)
+        b = random.randrange(a + 1, end)
+        assert start <= a < b <= end
+        intervals.append((a, b))
+    return intervals
+
+
+class IntervalSet(object):
+    """
+    Class representing a set of half-closed intervals.
+    """
+    def __init__(self, start, end, intervals):
+        self.start = start
+        self.end = end
+        self.intervals = sorted(intervals)
```

Comment:
this should probably sort only by start coordinate, not both start and end, as AFAICT that's not a requirement of returned variants or reads